nsxt/resource_nsxt_policy_qos_profile.go

```diff
@@ -0,0 +1,352 @@
+				Type:         schema.TypeInt,
+				Description:  "Class of service",
+				Optional:     true,
+				ValidateFunc: validation.IntBetween(0, 7),
+			},
+			"dscp_trusted": {
+				Type:        schema.TypeBool,
+				Description: "Trust mode for DSCP",
+				Optional:    true,
+				Default:     false,
+			},
+
+			"dscp_priority": {
+				Type:         schema.TypeInt,
+				Description:  "DSCP Priority",
+				Optional:     true,
+				ValidateFunc: validation.IntBetween(0, 63),
+			},
+			"ingress_rate_shaper":           getQosRateShaperSchema(ingressRateShaperIndex),
+			"ingress_broadcast_rate_shaper": getQosRateShaperSchema(ingressBroadcastRateShaperIndex),
+			"egress_rate_shaper":            getQosRateShaperSchema(egressRateShaperIndex),
+		},
+	}
+}
+
+func resourceNsxtPolicyQosProfileExists(id string, connector *client.RestConnector, isGlobalmodel bool) (bool, error) {
+	var err error
+	if isGlobalmodel {
+		client := gm_infra.NewDefaultQosProfilesClient(connector)
+		_, err = client.Get(id)
+	} else {
+		client := infra.NewDefaultQosProfilesClient(connector)
+		_, err = client.Get(id)
+	}
+	if err == nil {
+		return true, nil
+	}
+
+	if isNotFoundError(err) {
+		return false, nil
+	}
+
+	return false, logAPIError("Error retrieving resource", err)
+}
+
+func getPolicyQosRateShaperFromSchema(d *schema.ResourceData, index int) *data.StructValue {
+	scale := rateShaperScales[index]
+	schemaName := rateShaperSchemaNames[index]
+	resourceType := rateLimiterResourceTypes[index]
+	shaperConfs := d.Get(schemaName).([]interface{})
+	for _, shaperConf := range shaperConfs {
+		// only 1 is allowed for each type
+		shaperData := shaperConf.(map[string]interface{})
+		enabled := shaperData["enabled"].(bool)
+		averageBW := int64(shaperData[fmt.Sprintf("average_bw_%s", scale)].(int))
+		burstSize := int64(shaperData["burst_size"].(int))
+		peakBW := int64(shaperData[fmt.Sprintf("peak_bw_%s", scale)].(int))
+
+		shaper := model.IngressRateLimiter{
+			ResourceType:     resourceType,
+			Enabled:          &enabled,
+			BurstSize:        &burstSize,
+			AverageBandwidth: &averageBW,
+			PeakBandwidth:    &peakBW,
+		}
+		converter := bindings.NewTypeConverter()
+		converter.SetMode(bindings.REST)
+		dataValue, _ := converter.ConvertToVapi(shaper, model.IngressRateLimiterBindingType())
+		return dataValue.(*data.StructValue)
+	}
+
+	return nil
+}
+
+func setPolicyQosRateShaperInSchema(d *schema.ResourceData, shaperConf []*data.StructValue, index int) {
+	scale := rateShaperScales[index]
+	schemaName := rateShaperSchemaNames[index]
+	resourceType := rateLimiterResourceTypes[index]
+	var shapers []map[string]interface{}
+	converter := bindings.NewTypeConverter()
+	converter.SetMode(bindings.REST)
+	for _, dataShaper := range shaperConf {
+		dataValue, _ := converter.ConvertToGolang(dataShaper, model.IngressRateLimiterBindingType())
+		shaper := dataValue.(model.IngressRateLimiter)
+		if shaper.ResourceType == resourceType {
+			average := *shaper.AverageBandwidth
```

Comment:
are these guaranteed not to be nil?

Reply:
Done

Reply:
f.e. in gateway_prefix_list all prefix fields are optional

Reply:
In gateway_prefix_list there is no validation that optional fields are nil.

Reply:
It would be safer not to assume field being set in the struct, as we've seen NSX not specifying fields in response, which leads to these being set to nil in the SDK struct. 
Pointer assignment to schema map(like below) is safe, since terraform core handles nil use case:
elem["burst_size"] = shaper.BurstSize

But explicitly dereferencing pointer (like *shaper.BurstSize) should be nil-checked, unless we're 100% sure NSX will always sets these.

Reply:
ok. Thanks.